ntoskrnl/io/iomgr/deviface.c

```diff
@@ -412,6 +631,183 @@ IopOpenInterfaceKey(IN CONST GUID *InterfaceClassGuid,
+    PnpRegSzToString(Buffer, kvInfo->DataLength, &DeviceString.Length);
+    DeviceString.MaximumLength = (USHORT)kvInfo->DataLength;
+    DeviceString.Buffer = Buffer;
+
+    /* Separate symbolic link onto 4 parts:
+     * 1) prefix string (\??\ for kernel mode or \\?\ for user mode),
+     * 2) munged path string (like ##?#ACPI#PNP0501#1#{GUID}),
+     * 3) GUID string (the current GUID),
+     * 4) reference string (goes after GUID, starts with '\').
+     * 
+     * We need only prefix and reference strings.
+     */
+    Status = IopSeparateSymbolicLink(SymbolicLinkName,
+                                     &PrefixString,
+                                     NULL,
+                                     NULL,
+                                     &ReferenceString,
+                                     NULL,
+                                     NULL);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("Failed to separate '%wZ' with status 0x%08lx\n", SymbolicLinkName, Status);
+        goto Quit;
+    }
+
+    /* NOTE: Windows does some comparison of the string prefix here via RtlCompareMemory(),
+     * but in our case, this prefix is also passed to IopBuildSymolicLink() as well,
+     * so we don't need to compare anything here.
+     */
+
+    /* Replace all '\' separators with '#' in device string */
+    for (i = 0; i < DeviceString.Length; i++)
+    {
+        if (DeviceString.Buffer[i] == L'\\')
+        {
+            DeviceString.Buffer[i] = L'#';
+        }
+    }
+
+    DPRINT("Device string is '%wZ'\n", &DeviceString);
+
+    /* Build up new symbolic link with alias GUID */
+    Status = IopBuildSymbolicLink(&PrefixString,
+                                  &DeviceString,
+                                  &AliasGuidString,
+                                  &ReferenceString,
+                                  &AliasSymbolicLink);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("Failed to build alias symbolic link with status 0x%08lx\n", Status);
+        goto Quit;
+    }
+
+    /* Make sure that alias symbolic link key exists in registry */
+    Status = OpenRegistryHandlesFromSymbolicLink(&AliasSymbolicLink,
+                                                 KEY_READ,
+                                                 NULL,
+                                                 NULL,
+                                                 &AliasInstanceKey);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("Failed to open alias symbolic link key with status 0x%08lx\n", Status);
+        goto Quit;
+    }
+    ZwClose(AliasInstanceKey);
+
+    /* We're done */
+    DPRINT("IoGetDeviceInterfaceAlias() : alias symbolic link '%wZ'\n", &AliasSymbolicLink);
+    *AliasSymbolicLinkName = AliasSymbolicLink;
+    Status = STATUS_SUCCESS;
+
+Quit:
+    if (!NT_SUCCESS(Status))
+    {
+        if (AliasSymbolicLink.Buffer)
+            RtlFreeUnicodeString(&AliasSymbolicLink);
```

Comment:
In principle (I'm commenting here, but it's also valid for the other cases), freeing a unicode_string with `RtlFreeUnicodeString` requires it to have been allocated with `RtlCreateUnicodeString` (or at least, its buffers allocated with `RtlpAllocateStringMemory`); however it's not the case, you have used the ExAllocatePool(WithTag) functions. So it would be preferable to use the ExFreePool(WithTag) functions to free the buffers there, instead.

Reply:
OK, but need I zero the length in this case manually? I think yes, but I doubt a bit... :thinking:

Reply:
zero the length where?

Reply:
You already initialize their lengths to null, originally (at the top of this function).
Also, since these are variables that are only used locally in this function (with the exception of `AliasSymbolicLink` whose contents is returned by the pointer `AliasSymbolicLinkName`, but this is done only in success case, so we don't care about that here), they are not going to be used further elsewhere, so in this cleaning path, there is no need to reset their lengths to zero -- it won't matter at all.

Reply:
OK, indeed. Now I understand. Thank you. :sweat_smile: